File: src/handoff/daemon/failover.py
"""Failover policy: subscribes to RateLimitUpdated and decides when to hand off.

Per-project config lives in `<project>/.handoff/config.toml`:

    [failover]
    tokens_remaining_pct = 10.0     # trigger when tokens drop below this %
    tokens_remaining_abs = 1000     # OR when below this absolute count
    requests_remaining = 5          # OR when fewer than this many requests left
    chain = ["claude", "codex", "copilot"]
    auto_spawn = true
"""

from __future__ import annotations

import logging
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Optional

try:
    import tomllib  # py311+
except ImportError:  # pragma: no cover
    import tomli as tomllib  # type: ignore

from handoff.adapters import get_adapter
from handoff.context.engine import ContextEngine
from handoff.daemon.spawn import headless_spawn
from handoff.storage.db import Database

log = logging.getLogger("handoff.failover")


@dataclass
class FailoverConfig:
    tokens_remaining_pct: float = 10.0
    tokens_remaining_abs: Optional[int] = None
    requests_remaining: int = 5
    chain: tuple[str, ...] = ("claude", "codex", "copilot")
    auto_spawn: bool = True


def load_config(project_root: Path) -> FailoverConfig:
    cfg_path = project_root / ".handoff" / "config.toml"
    if not cfg_path.exists():
        return FailoverConfig()
    try:
        data = tomllib.loads(cfg_path.read_text())
    except Exception as e:  # noqa: BLE001
        log.warning("failed to parse %s: %s", cfg_path, e)
        return FailoverConfig()
    f = data.get("failover", {})
    chain = tuple(f.get("chain", FailoverConfig.chain))
    return FailoverConfig(
        tokens_remaining_pct=float(f.get("tokens_remaining_pct", 10.0)),
        tokens_remaining_abs=f.get("tokens_remaining_abs"),
        requests_remaining=int(f.get("requests_remaining", 5)),
        chain=chain,
        auto_spawn=bool(f.get("auto_spawn", True)),
    )


def should_trigger(
    cfg: FailoverConfig,
    *,
    tokens_remaining: Optional[int],
    requests_remaining: Optional[int],
    initial_tokens: Optional[int] = None,
) -> tuple[bool, str]:
    """Return (trigger, reason)."""
    if tokens_remaining is not None and cfg.tokens_remaining_abs is not None:
        if tokens_remaining < cfg.tokens_remaining_abs:
            return True, f"tokens_remaining={tokens_remaining} < abs={cfg.tokens_remaining_abs}"
    if (
        tokens_remaining is not None
        and initial_tokens is not None
        and initial_tokens > 0
    ):
        pct = 100.0 * tokens_remaining / initial_tokens
        if pct < cfg.tokens_remaining_pct:
            return True, f"tokens_remaining={tokens_remaining} ({pct:.1f}% < {cfg.tokens_remaining_pct}%)"
    if requests_remaining is not None and requests_remaining < cfg.requests_remaining:
        return True, f"requests_remaining={requests_remaining} < {cfg.requests_remaining}"
    return False, ""


class FailoverPolicy:
    """Subscribes to RateLimitUpdated; spawns next agent in chain on trigger.

    De-duplication: once a (from_agent_id, to_kind) handoff fires, that pair
    is not retried until the failing agent's row is replaced (i.e. ended_at
    is set).
    """

    def __init__(self, db: Database):
        self.db = db
        self._fired: set[tuple[int, str]] = set()
        self._initial_tokens: dict[int, int] = {}

    def on_rate_limit(self, payload: dict[str, Any]) -> None:
        agent_id = payload.get("agent_id")
        if agent_id is None:
            return
        tokens = payload.get("tokens_remaining")
        requests = payload.get("requests_remaining")

        # Track the highest tokens_remaining we've ever seen for this agent
        # as a stand-in for the initial budget (no provider gives us limit).
        if tokens is not None:
            prev = self._initial_tokens.get(agent_id, 0)
            if tokens > prev:
                self._initial_tokens[agent_id] = tokens

        project_id = self.db.project_id_for_agent(agent_id)
        if project_id is None:
            return
        root_str = self.db.project_root(project_id)
        if not root_str:
            return
        root = Path(root_str)
        cfg = load_config(root)

        trigger, reason = should_trigger(
            cfg,
            tokens_remaining=tokens,
            requests_remaining=requests,
            initial_tokens=self._initial_tokens.get(agent_id),
        )
        if not trigger:
            return

        next_kind = self._pick_next(cfg.chain, payload.get("kind"))
        if next_kind is None:
            log.info("no failover candidate for kind=%s in chain", payload.get("kind"))
            return

        key = (agent_id, next_kind)
        if key in self._fired:
            return
        self._fired.add(key)

        log.warning(
            "failover triggered: agent_id=%s kind=%s reason=%s -> %s",
            agent_id, payload.get("kind"), reason, next_kind,
        )
        self.execute(
            from_agent_id=agent_id,
            to_kind=next_kind,
            project_root=root,
            project_id=project_id,
            reason=reason,
            auto_spawn=cfg.auto_spawn,
        )

    def execute(
        self,
        *,
        from_agent_id: Optional[int],
        to_kind: str,
        project_root: Path,
        project_id: int,
        reason: str,
        auto_spawn: bool = True,
    ) -> dict:
        engine = ContextEngine(project_root)
        snapshot = engine.snapshot(
            note=f"Failover from agent_id={from_agent_id}: {reason}"
        )

        new_agent_id: Optional[int] = None
        spawn_pid: Optional[int] = None
        if auto_spawn:
            try:
                proc = headless_spawn(
                    to_kind,
                    project_root,
                    prompt=f"Resuming work. Read this snapshot first: {snapshot}",
                )
                if proc is not None:
                    spawn_pid = proc.pid
                    new_agent_id = self.db.insert_agent(
                        project_id=project_id,
                        kind=to_kind,
                        pid=spawn_pid,
                        spawned_by="handoff",
                    )
            except Exception as e:  # noqa: BLE001
                log.exception("failover spawn failed: %s", e)

        handoff_id = self.db.insert_handoff(
            from_agent_id=from_agent_id,
            to_agent_id=new_agent_id,
            reason=reason,
            snapshot_path=str(snapshot),
        )
        return {
            "handoff_id": handoff_id,
            "to_agent_id": new_agent_id,
            "to_pid": spawn_pid,
            "snapshot_path": str(snapshot),
        }

    @staticmethod
    def _pick_next(chain: tuple[str, ...], current: Optional[str]) -> Optional[str]:
        for k in chain:
            if k != current:
                return k
        return None

File: rust/crates/daemon/src/failover.rs
//! Failover engine: subscribes to ingest events on an mpsc channel, decides
//! whether to fire a handoff, and (when `auto_spawn=true`) launches the next
//! agent in the chain.

use std::collections::{HashMap, HashSet};
use std::path::Path;
use std::sync::Arc;

use anyhow::Result;
use handoff_context::ContextEngine;
use handoff_critic::CriticRunner;
use handoff_policy::{load as load_policy, pick_next, should_trigger};
use handoff_storage::Database;
use tokio::sync::{mpsc, Mutex};
use tracing::{info, warn};

use crate::spawn::headless_spawn;

#[derive(Debug, Clone)]
pub struct RateEvent {
    pub agent_id: i64,
    pub kind: String,
    pub tokens_remaining: Option<i64>,
    pub requests_remaining: Option<i64>,
}

#[derive(Debug, Clone)]
pub struct HandoffOutcome {
    pub handoff_id: i64,
    pub to_agent_id: Option<i64>,
    pub to_pid: Option<u32>,
    pub snapshot_path: String,
}

/// Default channel capacity. Tune in production.
pub const CHANNEL_CAP: usize = 256;

pub fn channel() -> (mpsc::Sender<RateEvent>, mpsc::Receiver<RateEvent>) {
    mpsc::channel(CHANNEL_CAP)
}

pub struct FailoverEngine {
    db: Arc<Database>,
    proxy_url: String,
    /// (from_agent_id, to_kind) pairs that already fired.
    fired: Mutex<HashSet<(i64, String)>>,
    /// Highest `tokens_remaining` ever seen per agent — proxy for "initial budget".
    initial_tokens: Mutex<HashMap<i64, i64>>,
}

struct ExecuteRequest<'a> {
    from_agent_id: Option<i64>,
    to_kind: &'a str,
    project_root: &'a Path,
    project_id: i64,
    reason: &'a str,
    auto_spawn: bool,
    summarize: bool,
    use_worktree: bool,
}

impl FailoverEngine {
    pub fn new(db: Arc<Database>, proxy_url: String) -> Self {
        Self {
            db,
            proxy_url,
            fired: Mutex::new(HashSet::new()),
            initial_tokens: Mutex::new(HashMap::new()),
        }
    }

    /// Spawn the engine task and return its sender. Drop the sender to stop it.
    pub fn spawn(self: Arc<Self>) -> mpsc::Sender<RateEvent> {
        let (tx, mut rx) = channel();
        let me = self.clone();
        tokio::spawn(async move {
            while let Some(ev) = rx.recv().await {
                if let Err(e) = me.handle(ev).await {
                    warn!("failover handle error: {e}");
                }
            }
        });
        tx
    }

    pub async fn handle(&self, ev: RateEvent) -> Result<()> {
        if let Some(tok) = ev.tokens_remaining {
            let mut m = self.initial_tokens.lock().await;
            let prev = m.get(&ev.agent_id).copied().unwrap_or(0);
            if tok > prev {
                m.insert(ev.agent_id, tok);
            }
        }
        let project_id = match self.db.project_id_for_agent(ev.agent_id)? {
            Some(p) => p,
            None => return Ok(()),
        };
        let root = match self.db.project_root(project_id)? {
            Some(r) => r,
            None => return Ok(()),
        };
        let root = std::path::PathBuf::from(root);
        let policy = load_policy(&root).unwrap_or_default();
        let initial = self.initial_tokens.lock().await.get(&ev.agent_id).copied();
        let trigger = should_trigger(
            &policy.failover,
            ev.tokens_remaining,
            ev.requests_remaining,
            initial,
        );
        if !trigger.fired {
            return Ok(());
        }

        let samples: Vec<handoff_policy::RateSampleInput> = policy
            .failover
            .chain
            .iter()
            .filter_map(|kind| {
                self.db
                    .latest_rate_sample_for_kind(kind)
                    .ok()
                    .flatten()
                    .map(|s| handoff_policy::RateSampleInput {
                        kind: kind.clone(),
                        tokens_remaining: s.tokens_remaining.unwrap_or(0),
                        tokens_reset_at: s.tokens_reset_at,
                    })
            })
            .collect();

        let next_kind = match pick_next(&policy.failover.chain, &ev.kind, &samples) {
            Some(k) => k.to_string(),
            None => return Ok(()),
        };
        let key = (ev.agent_id, next_kind.clone());
        {
            let mut fired = self.fired.lock().await;
            if fired.contains(&key) {
                return Ok(());
            }
            fired.insert(key);
        }
        info!(
            "failover triggered: agent_id={} kind={} reason={} -> {}",
            ev.agent_id, ev.kind, trigger.reason, next_kind,
        );
        self.execute_inner(ExecuteRequest {
            from_agent_id: Some(ev.agent_id),
            to_kind: &next_kind,
            project_root: &root,
            project_id,
            reason: &trigger.reason,
            auto_spawn: policy.failover.auto_spawn,
            summarize: policy.failover.summarize,
            use_worktree: policy.failover.use_worktree,
        })
        .await?;
        Ok(())
    }

    /// Manual / RPC entry point. Returns the `HandoffOutcome` so the caller
    /// can render it.
    pub async fn execute(
        &self,
        from_agent_id: Option<i64>,
        to_kind: &str,
        project_root: &Path,
        project_id: i64,
        reason: &str,
        auto_spawn: bool,
    ) -> Result<HandoffOutcome> {
        let policy = load_policy(project_root).unwrap_or_default();
        self.execute_inner(ExecuteRequest {
            from_agent_id,
            to_kind,
            project_root,
            project_id,
            reason,
            auto_spawn,
            summarize: policy.failover.summarize,
            use_worktree: policy.failover.use_worktree,
        })
        .await
    }

    async fn execute_inner(&self, req: ExecuteRequest<'_>) -> Result<HandoffOutcome> {
        let ExecuteRequest {
            from_agent_id,
            to_kind,
            project_root,
            project_id,
            reason,
            auto_spawn,
            summarize,
            use_worktree,
        } = req;

        let engine = ContextEngine::new(project_root);
        let (mut snap, _md_path) = engine.snapshot(reason)?;
        // Ask the critic agent for a focused brief if configured. Failure
        // is non-fatal — we fall back to the verbatim brain dump.
        if summarize {
            let policy = handoff_policy::load(project_root).unwrap_or_default();
            if let Ok(runner) = CriticRunner::new(project_root) {
                let runner = runner
                    .with_agents(&policy.critic.worker_agent, &policy.critic.summarizer_agent)
                    .with_proxy(Some(self.proxy_url.clone()));
                match runner.summarize_for_handoff(reason).await {
                    Ok(brief) => snap.critic_brief = Some(brief),
                    Err(e) => warn!("summarizer failed; verbatim fallback: {e}"),
                }
            }
        }
        let (snap_path, _) = engine.write_snapshot(&snap)?;
        let snap_path_str = snap_path.display().to_string();

        let mut new_agent_id: Option<i64> = None;
        let mut new_pid: Option<u32> = None;
        if auto_spawn {
            // Insert agent record first to get an ID for the worktree
            let aid = self.db.insert_agent(
                project_id, to_kind, None, // PID unknown yet
                "handoff",
            )?;
            new_agent_id = Some(aid);

            let prompt = format!("Resuming work. Read this snapshot first: {}", snap_path_str);
            match headless_spawn(
                aid,
                to_kind,
                project_root,
                &prompt,
                &self.proxy_url,
                use_worktree,
            )
            .await
            {
                Ok(Some(child)) => {
                    new_pid = child.id();
                    if let Some(pid) = new_pid {
                        self.db.update_agent_pid(aid, pid as i64)?;
                    }
                }
                Ok(None) => {
                    info!("no headless form for kind={to_kind}; skipping spawn");
                }
                Err(e) => warn!("failover spawn failed: {e}"),
            }
        }

        let handoff_id =
            self.db
                .insert_handoff(from_agent_id, new_agent_id, reason, Some(&snap_path_str))?;
        Ok(HandoffOutcome {
            handoff_id,
            to_agent_id: new_agent_id,
            to_pid: new_pid,
            snapshot_path: snap_path_str,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn dedup_per_pair() {
        let tmp = tempfile::tempdir().unwrap();
        let db_path = tmp.path().join("t.db");
        let db = Arc::new(Database::open(&db_path).unwrap());
        let pid = db
            .upsert_project(&tmp.path().display().to_string())
            .unwrap();

        // Scaffold .handoff/ in the project root so snapshot/write succeed.
        std::fs::create_dir_all(tmp.path().join(".handoff").join("scratch")).unwrap();
        std::fs::write(
            tmp.path().join(".handoff").join("config.toml"),
            r#"[failover]
requests_remaining = 100
auto_spawn = false
chain = ["codex", "claude"]
"#,
        )
        .unwrap();
        std::fs::write(tmp.path().join(".handoff").join("brain.md"), "# brain").unwrap();

        let aid = db.insert_agent(pid, "claude", Some(999), "user").unwrap();
        let engine = Arc::new(FailoverEngine::new(
            db.clone(),
            "http://127.0.0.1:8080".into(),
        ));

        let ev = RateEvent {
            agent_id: aid,
            kind: "claude".into(),
            tokens_remaining: None,
            requests_remaining: Some(1),
        };
        engine.handle(ev.clone()).await.unwrap();
        engine.handle(ev).await.unwrap();

        // Only one handoff row should exist; query via a fresh connection.
        // Drop the in-memory db's reference so its Mutex doesn't fight us.
        drop(engine);
        drop(db);
        let n = count_handoffs(&db_path, aid);
        assert_eq!(n, 1);
    }

    fn count_handoffs(path: &std::path::Path, aid: i64) -> i64 {
        let conn = rusqlite::Connection::open(path).unwrap();
        conn.query_row::<i64, _, _>(
            "SELECT COUNT(*) FROM handoffs WHERE from_agent_id = ?1",
            rusqlite::params![aid],
            |r| r.get(0),
        )
        .unwrap()
    }
}
